Enabled support for partial derivatives in as_finite_diff

arihantparsoya opened the pull request:
This is fix for #11007 . Previously the function `as_finite_diff` evaluated only the function of the differential equation passed to it and ignored if the expression was differentiated multiple times.
#### Output before this PR:

``` Python
>>> f = Function('f')(x, y)
>>> d2fdxdy = f.diff(x).diff(y)
>>> as_finite_diff(d2fdxdy, wrt=x)
-f(x - 1/2, y) + f(x + 1/2, y)
```
#### Output after this PR:

``` Python
>>> f = Function('f')(x, y)
>>> d2fdxdy = f.diff(x).diff(y)
>>> pprint(as_finite_diff(d2fdxdy, wrt=x))
  ∂                   ∂                
- ──(f(x - 1/2, y)) + ──(f(x + 1/2, y))
  ∂y                  ∂y               
```

arihantparsoya committed:
Merge remote-tracking branch 'sympy/master'

arihantparsoya committed:
Merge remote-tracking branch 'sympy/master'

arihantparsoya committed:
Merge remote-tracking branch 'sympy/master'

arihantparsoya committed:
Merge remote-tracking branch 'sympy/master'

arihantparsoya committed:
Merge remote-tracking branch 'sympy/master'

arihantparsoya committed:
Merge remote-tracking branch 'sympy/master'

arihantparsoya committed:
made finite_diff to account for functions with multiple derivatives

tests are added

parsoyaarihant commented:
@bjodah , @asmeurer , can you review this?

arihantparsoya committed:
changed the doctest

parsoyaarihant commented:
@bjodah  , @asmeurer  , can you review this?

bjodah commented:
Thank you for this patch @parsoyaarihant 
I think the suggested changes looks good and adequately addresses the issue. +1 for merging.

mloubout commented:
Can the new output be passed back to as_finite_diff in order to get the full discretization of diff(x,y) wrt=(x,y) : 
as_finite_diff(as_finite_diff(d2fdxdy, wrt=x),wrt=y)

bjodah commented:
@mloubout agreed, `as_finite_diff` should eventually work for non-`Derivative` instances, e.g. sums of derivatives etc.
I would suggest opening a different issue for that, unless @parsoyaarihant want to work on that immediately in this PR?

arihantparsoya committed:
enabled support for is_Add expressions

parsoyaarihant commented:
I improved the function to account for some non-derivative expression. Please check my last commit and comment.

bjodah commented:
I'm not quite sure 4283cff is the best approach, special-casing with `is_Add` might be fine, but then we need to add other special cases as well, e.g. `is_Mul`. And if we support `is_Add` we should definitely support `as_finit_diff(dfdx + 1, wrt=x)`, ... and arbitrarily nested expressions.

One approach that might be viable would be to have `as_finit_diff` apply itself recursively on all `.args` (if not `is_Derivative` and not `is_Atom`).

parsoyaarihant commented:
Doing recursively on all .args is the right approach. I have a better idea for this , I will push commit for it soon.

arihantparsoya committed:
enabled is_Mul expressions in as_finite_diff

used chain rule to return the final result. If the subexpression in
derivatives.args cannot be evaluated (such as integers) then
subexpression is multiplied as it is to the final result.

parsoyaarihant commented:
@bjodah , I have made the necessary changes to support for `is_Add` and `is_Mul` type expressions. 

- used chain rule to return the final result in `is_Mul` type expressions. If the subexpression in derivatives.args cannot be evaluated (such as integers) then subexpression is multiplied as it is to the final result.

#### Sample Program:

```python
>>> from sympy import *
>>> from sympy.abc import x,y
>>> f = Function('f')(x, y)
>>> g = Function('g')(x, y)
>>> d2fdxdy = f.diff(x).diff(y)
>>> dfgdx  = (f*g).diff(x)
>>> pprint(as_finite_diff(dfgdx, wrt=x))
(-f(x - 1/2, y) + f(x + 1/2, y))⋅g(x, y) + (-g(x - 1/2, y) + g(x + 1/2, y))⋅f(
x, y)
>>> pprint(as_finite_diff(as_finite_diff(d2fdxdy, wrt=x),wrt=y))
f(x - 1/2, y - 1/2) - f(x - 1/2, y + 1/2) - f(x + 1/2, y - 1/2) + f(x + 1/2, y
 + 1/2)
>>> pprint(as_finite_diff(dfgdx + 1, wrt=x))
(-f(x - 1/2, y) + f(x + 1/2, y))⋅g(x, y) + (-g(x - 1/2, y) + g(x + 1/2, y))⋅f(
x, y) + 1
>>> pprint(as_finite_diff(5*(dfgdx) + 1, wrt=x))
5⋅(-f(x - 1/2, y) + f(x + 1/2, y))⋅g(x, y) + 5⋅(-g(x - 1/2, y) + g(x + 1/2, y)
)⋅f(x, y) + 1
```

bjodah commented:
@parsoyaarihant thank you for working on this.
I still think we can be more general here. Imagine that a user supplies an expression containing some `Derivative` instances, we do not know what this expression is, all we want to do is to traverse the tree and replace all Derivative instances containing our `wrt` with a finite difference approximations. `is_Mul` would miss e.g. `as_finite_diff(1/d2fdxdy, wrt=x)` etc.

I would do something along the lines of:

```
if derivative.is_Derivative:
    pass
elif derivative.is_Atom:
    return derivative
else:
    return derivative.fromiter([as_finite_diff(arg, wrt=wrt, ...) for arg in self.args], **derivative.assumptions0)
```
note that this completely untested (and I may very well have overseen something). But in theory that kind of approach should be able to support things such as `as_finite_diff(exp(d2fdxdy), wrt=x))`.

mloubout commented:
I also still think we would need a flag spefying if the input hqs to be expanded (or other) as a lot of numerical methods require it to be onsidered as a single function to evaluate on the finite differencfe grid

bjodah commented:
@mloubout that would surely be valuable, but don't you think it can be added in a subsequent PR? or would that be too late?

mloubout commented:
@bjodah  It depends, I am not sure how hard it is to create a flag something like 'noeval' to only output finite_diif for f*g as :
(f*g)(x+1/2) - (f*g)(x-1/2) 
like if it was a single function. I'll let you decide as you do the heavy lifting

bjodah commented:
@mloubout that shouldn't be too hard (using e.g. `expr.subs({f(x)*g(x): h(x)})`), however I am not convinced we need to teach as_finite_diff about it (we can inform the user in the docstring how to achieve this though). Let's keep the scope of this PR limited unless it can be shown that it would hinder future enhancements.

mloubout commented:
OK, then let's just focus on the pure derivatives as you stated above to  make it robust to any input and put this special case in the docstring.

arihantparsoya committed:
making as_finite_diff more robust

parsoyaarihant commented:
@bjodah , @mloubout , I made changes to apply `as_finite_diff` to all derivative instances in the expression tree. 

#### Sample Program:
```Python
>>> from sympy import *
>>> from sympy.abc inmport x,y
  File "<stdin>", line 1
    from sympy.abc inmport x,y
                         ^
SyntaxError: invalid syntax
>>> from sympy.abc import x,y
>>> f = Function('f')(x, y)
>>> g = Function('g')(x, y)
>>> d2fdxdy = f.diff(x).diff(y)
>>> pprint(as_finite_diff(exp(d2fdxdy), wrt=x))
   ∂                   ∂                
 - ──(f(x - 1/2, y)) + ──(f(x + 1/2, y))
   ∂y                  ∂y               
ℯ                                       
>>> pprint(as_finite_diff(1/(d2fdxdy), wrt=x))
                   1                   
───────────────────────────────────────
  ∂                   ∂                
- ──(f(x - 1/2, y)) + ──(f(x + 1/2, y))
  ∂y                  ∂y               
>>> dfgdx  = (f*g).diff(x)
>>> pprint(as_finite_diff(dfgdx, wrt=x))
(-f(x - 1/2, y) + f(x + 1/2, y))⋅g(x, y) + (-g(x - 1/2, y) + g(x + 1/2, y))⋅f(
x, y)
```

mloubout commented:
What does it do for 


`>>> dfgdx  = (f/g).diff(x)`

parsoyaarihant commented:
@mloubout : 
```python
>>> from sympy import *
>>> from sympy.abc import x,y
>>> f = Function('f')(x, y)
>>> g = Function('g')(x, y)
>>> dfgdx = (f/g).diff(x)
>>> pprint(as_finite_diff(dfgdx, wrt=x))
-f(x - 1/2, y) + f(x + 1/2, y)   (-g(x - 1/2, y) + g(x + 1/2, y))⋅f(x, y)
────────────────────────────── - ────────────────────────────────────────
           g(x, y)                                2                      
                                                 g (x, y) 
```

mloubout commented:
@parsoyaarihant cooool

parsoyaarihant commented:
@smichr can you review this?

bjodah commented:
@parsoyaarihant I like your use of replace.
We should add some more tests of some nested expressions.
Trying out your branch I found another bug (present even before this PR):
```
>>> dfdx = f(x).diff(x)
>>> as_finite_diff(dfdx, wrt=x)
-f(x - 1/2) + f(x + 1/2)
>>> as_finite_diff(dfdx, wrt=y)
f(x)
```
(where we would have expected to just get our dfdx back), since this PR enables `as_finite_diff` to operate on aribtrary expressions, we should fix that too (by inspecting the `.variables` attr of our `Derivative`)

mloubout commented:
@bjodah @parsoyaarihant 
On the previous example 
`>>> as_finite_diff(dfdx, wrt=y)
f(x)
`
you should actually get back 0 as f is independent of y

bjodah commented:
@mloubout I don't think so. `as_finite_diff` does not differentiate the expression (it merely expresses existing derivatives in terms of a finite difference approximation).

mloubout commented:
@bjodah yes but it should output
f(x,y-1/2) - f(x,y+1/2)
but if f only depends on x, no matter what y is it should be zero. This not outputting 0 is hugely problematic in a lot of levels. Variable independence is extremely important. If you do not , then all the differentiation becomes irrelevant as you should have then
`as_finite_diff(dfdx, wrt=y) 
(y(x-1/2) - y(x+1/2)) (f(x(y-1/2))- f(x(x+1/2)))`
It is absolutely necessary that is a function is independent of the variable you are differentiating against you should get 0 the same way you get constant.diff(x) = 0 no matter what constant is

bjodah commented:
@mloubout sorry, I have some trouble following your argument. I'm not even sure if we are talking about the same thing, here is what I expect `as_finite_diff` to do:
```
>>> as_finite_diff(1) == 1
True
>>> as_finite_diff(1 + f(x).diff(x), wrt=x) == 1 + f(x + S(1)/2) - f(x - S(1)/2)
True
>>> expr = f(x).diff(x) + g(y).diff(y)
>>> as_finite_diff(expr, wrt=y) == f(x).diff(x) + as_finite_diff(g(y).diff(y), wrt=y)
True
```
I can't see how this would be violating "variable independence".

mloubout commented:
if f(y) = 1 (so constant)
as_finite_diff(f,wrt=y)   = f(y-1/2) - f(y+1/2) = 1 - 1 = 0
if f(x,y) = h(x)  (so constant in y)
as_finite_diff(f,wrt=y)   = f(x,y-1/2) - f(x,y+1/2) = h(x) - h(x) = 0

bjodah commented:
@mloubout no, `as_finite_diff` operates on `Derivative` instances, not `Function` instances.

mloubout commented:
`symbols('s)
s.diff(x)
0`
It doesn't matter what you have, a derivative, function, symbol .... if it does not depends on the variable inside diff() or wrt = the result IS 0 always.

bjodah commented:
I'm not arguing about basic calculus, `as_finite_diff` is not an operator in that sense, look at it as something like `replace` in SymPy, i.e. `as_finite_diff` should never differentiate an expression.

mloubout commented:
as_finit_diff compute the taylor expansion by evaluating the input f at different grid points.  If it does not output 0 for a constant there is a big problem

asmeurer commented on sympy/calculus/tests/test_finite_diff.py:
```diff
@@ -175,3 +175,7 @@ def test_as_finite_diff():
+    #issue 11007
+    f = Function('f')(x, h)
+    d2fdxdh = f.diff(x).diff(h)
+    assert str(as_finite_diff(d2fdxdh, wrt=x)) == '-Derivative(f(x - 1/2, h), h) + Derivative(f(x + 1/2, h), h)'
```
Please don't test strings. You should only test strings in printing tests (which this is not).

asmeurer commented:
So I guess this can go either way. Right now, in master, `as_finite_diff` assumes the entire expression is a Derivative  (`as_finite_diff(1, x)` raises an exception for instance). So what exactly is the function *supposed* to do? Is it supposed to treat the entire expression as a derivative, or is it supposed to replace all derivatives in subexpressions? What are the use-cases for this function? What was the use-case for when it was originally added?

bjodah commented:
Originally it was supposed to replace a `Derivative` instance of an ordinary derivative with a finite difference approximation. It worked for ordinary derivatives, this PR would extend it to work also for partial derivatives, and as a bonus it would also extend it to operate on arbitrary expressions containing `Derivative` instances. The use case when I originally wrote the function was code generation, and judging by the references in the original issue, still is. I am -1 to extending `as_finite_diff` to perform any differentiation (that would be a breaking change needing a deprecation cycle). If there is interest for such a function we should simply add another one (which I am all for, but in my book, that is another issue).

There is one thing which we haven't discussed yet, and that it how to extend the `wrt` kwarg in light of the added capabilities of `as_finite_diff`.

arihantparsoya committed:
Added tests

asmeurer commented:
So what would a simple code generation workflow look like, for an example where you have an expression containing a Derivative instance?

arihantparsoya committed:
Typo

parsoyaarihant commented:
[Here](http://docs.sympy.org/dev/tutorial/calculus.html) in documentation, it specifies that the function acts on derivative instance. If the input expression is instance of derivative then, then we expect the right output anyway with this PR. I think its an additional feature to apply `as_finite_diff` function on all subexpression derivative instance in the expression since it also enables non-support mode. Previously, the function couldnt act on expressions such as multiplication and division on 2 functions but now it can:
```python
>>> from sympy import *
>>> from sympy.abc import x,y
>>> f = Function('f')(x, y)
>>> g = Function('g')(x, y)
>>> dfgdx = (f/g).diff(x)
>>> pprint(as_finite_diff(dfgdx, wrt=x))
-f(x - 1/2, y) + f(x + 1/2, y)   (-g(x - 1/2, y) + g(x + 1/2, y))⋅f(x, y)
────────────────────────────── - ────────────────────────────────────────
           g(x, y)                                2                      
                                                 g (x, y) 
```

arihantparsoya committed:
Made correction for derivative instances

parsoyaarihant commented:
@bjodah , @asmeurer @mloubout , I made necessary changes in the PR:

#### Sample Program:
```python 
>>> from sympy import *
>>> from sympy.abc import x,y,z
>>> from sympy import I
>>> f = Function('f')(x, y)
>>> g = Function('g')(x,y)
>>> d2fdxdy = (f*g).diff(x)
>>> dfgdx = (f/g).diff(x)
>>> pprint(as_finite_diff(dfgdx + exp(x), wrt=x))
-f(x - 1/2, y) + f(x + 1/2, y)   (-g(x - 1/2, y) + g(x + 1/2, y))⋅f(x, y)
────────────────────────────── - ────────────────────────────────────────
           g(x, y)                                2                      
                                                 g (x, y)                
>>> pprint(as_finite_diff(d2fdxdy, wrt=x))
(-f(x - 1/2, y) + f(x + 1/2, y))⋅g(x, y) + (-g(x - 1/2, y) + g(x + 1/2, y))⋅f(
x, y)
>>> pprint(as_finite_diff(exp(d2fdxdy), wrt=x))
0
```

* Please note that only the subexpressions having `wrt` is being evaluated.

bjodah commented:
Sorry, I don't understand the value of 51c3a73 now you are special casing `is_Add` again?

parsoyaarihant commented:
Yes, I am thinking to create a extra flag(work in progress) if the user wants to apply as_finite_diff on all derivative instances or use the expression as insisted by @mloubout .

arihantparsoya committed:
added an additional flag applyall

parsoyaarihant commented:
I have created an additional flag called `applyall` for `as_finite_diff`. When `applyall=True` it uses the `replace` method to replace all derivative instances in the expression. When `applyall=False`, the function treats the whole input expression as a derivative instance. 

* Note: `applyall` is set default to `False`.

#### Sample Program:
``` Python
>>> from sympy import *
>>> from sympy.abc import x,y
>>> f = Function('f')(x, y)
>>> g = Function('g')(x,y)
>>> dfgdx = (f/g).diff(x)
>>> pprint(as_finite_diff(dfgdx, wrt=x))
-f(x - 1/2, y) + f(x + 1/2, y)   (-g(x - 1/2, y) + g(x + 1/2, y))⋅f(x, y)
────────────────────────────── - ────────────────────────────────────────
           g(x, y)                                2                      
                                                 g (x, y)                
>>> 
>>> dfgdx = (f*g).diff(x)
>>> pprint(as_finite_diff(dfgdx, wrt=x))
(-f(x - 1/2, y) + f(x + 1/2, y))⋅g(x, y) + (-g(x - 1/2, y) + g(x + 1/2, y))⋅f(
x, y)
>>> pprint(as_finite_diff(exp(dfgdx), wrt=x))
0
>>> pprint(as_finite_diff(exp(dfgdx), wrt=x, applyall=True))
 (-f(x - 1/2, y) + f(x + 1/2, y))⋅g(x, y) + (-g(x - 1/2, y) + g(x + 1/2, y))⋅f (x, y)
ℯ                                                                            
```

mloubout commented:
Sorry for complaining again, but this still does not follow the mathematics 
`pprint(as_finite_diff(dfgdx + exp(x), wrt=x))`
why does it treat exp(x) as a constant? It explicitly depends on x.
` pprint(as_finite_diff(exp(dfgdx), wrt=x, applyall=True))`
I do not understand the output. 

I still think the main problem is that everything is based on derivative instance, while Finite Differences are supposed to act on functions

bjodah commented:
Perhaps `as_finite_diff` was a poorly chosen name (it was originally supposed to be a method in `Derivative`).

So perhaps we want a new function `diff_finite` (just a suggestion, there are probably better names) which actually performs differentiation along the lines of @mloubout suggestions.

We could then rename `as_finite_diff` to `Derivative_as_finite_diff` and keep a (deprecated) alias around for a deprecation cycle. What are your feelings to this?

parsoyaarihant commented:
@bjodah , I think you are right. As, @mloubout pointed out that the problem is that the `as_finite_diff` is made to act only on Deriavtive instances hence its better to have another function which acts on functions in general. But, what about the current implementations? Should I revert back the changes and start fresh on an new PR for the new function?

asmeurer commented:
I agree that if the function is only supposed to act on Derivative instances, then it should be a method of Derivative.

bjodah commented:
The [original motivation](https://github.com/sympy/sympy/pull/7382#issuecomment-40838031) for placing it in `.calculus.finite_diff` was to avoid complicated logic in core.

parsoyaarihant commented:
@bjodah , @asmeurer , @mloubout , I have opened PR #11242 for addressing the original issue since there were many irrelevant commits in this PR. Please share your ideas on the new function as @bjodah and @mloubout suggested, I will begin working in it soon.

parsoyaarihant commented:
@mloubout , can you describe the worflow the the new function `finite_diff` ? To apply finite difference approximation to a general expression, what should the function exactly do when the input is not an instance of  `Derivative` ?

mloubout commented:
@parsoyaarihant It should be possible to apply it to :
- Derivatives instances as it is currently
- Functions : Evaluate the function at finite difference points 
`f(x - 1/2, y) + f(x + 1/2, y)` if thew input is `f(x,y)` for example
- Constants : as to return zeros

czgdp1807 commented:
On the latest `master`, 
```python
>>> f = Function('f')(x, y)
>>> d2fdxdy = f.diff(x).diff(y)
>>> d2fdxdy.as_finite_difference(wrt=x)
  ∂                   ∂                
- ──(f(x - 1/2, y)) + ──(f(x + 1/2, y))
  ∂y                  ∂y               
>>> 
```
Closing this. Thanks for the contributions. Please let us know if there is a need to re-open this.
